packages/models-library/src/models_library/services.py

```diff
@@ -115,7 +116,29 @@ class Config:
         frozen = True
 
 
-RunID = UUID
+class RunID(str):
+    """
+    Used to assign a unique identifier to the run of a service.
+
+    Example usage:
+    The dynamic-sidecar uses this to distinguish between current
+    and old volumes for different runs.
+    Avoids overwriting data that left dropped on the node (due to an error)
+    and gives the osparc-agent an opportunity to back it up.
+    """
```

Comment:
I personally do not understand this much.

So you create a run identifier, yes I could say as much from the name of the variable (which is good).

But actually you only use this for the sydecar volume? nothing else? You wrote above the 2 sidecars with the same RunID cannot happen cause there cannot be 2 sidecars with the same name. So is the RunID also in the name of the sidecar or this is only due to the nodeID?

What is unclear, is that the RunID in itself could have collided (because of the second accuracy). Now it is not the case anymore since you combine it with a UUID. now I do have a slightly different question. Are you living in the Matrix? I was just wondering on the usefulness of the seconds in int... especially for debugging that is not super easy to use ;)


Anyway, let's go with this.

Reply:
The volumes are a particular type of resource, we want them to have the following properties:
- they need to have unique names for each run (so using the node_id would not be enough)
- they need to be picked up by a third party (osparc-agent) and it needs to distinguish between different volumes coming from the same sidecar since it backs up the data from the volumes in the following path in s3 `project_id/node_id/run_id/volume_name`.
- the issue with the `run_id` here is that it has to be not only unique but easily sortable so we can tell which one is the last updated. In s3 there are no such things as folders and we cannot rely on the dates of the files, since when we have thousands of files it is not feasible to search them to figure out their last modified date

So the `run_id` is generated when the service is accepted by the director-v2 and is started.

Finally the integer timestamp part of the run_id allows to easily sort all the entries inside each service's backed up data and figure out which is the latest uploaded version.